numpyro/infer/hmc_gibbs.py

```diff
@@ -422,8 +435,16 @@ def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
         u_new = {}
         for name in gibbs_sites:
             size, subsample_size = plate_sizes[name]
-            rng_key, subkey = random.split(rng_key)
-            u_new[name] = random.choice(subkey, size, (subsample_size,), replace=False)
+            num_block = num_blocks[name]
+            rng_key, subkey, block_key = random.split(rng_key, 3)
+            block_size = subsample_size // num_block
+
+            chosen_block = random.randint(block_key, shape=(), minval=0, maxval=num_block)
+            new_idx = random.randint(subkey, minval=0, maxval=size, shape=(subsample_size,))
+            block_mask = (jnp.arange(subsample_size) // block_size == chosen_block).astype(int)
```

Comment:
nit: no need to use `astype` here, then you can use `u_new[name] = jnp.where(block_mask, new_idx, gibbs_sites[name])`

Reply:
With integer arithmetics the control flow can be executed using fully data parallel operations. So while it comprises readability it's measurably faster when run on parallel hardware.

Reply:
Oh, it is good to know. Make sense to me but I need to double check. In the mean time, you can define `rest_mask = 1 - block_mask` I guess.

Reply:
Just for reference,
```python
import jax.numpy as jnp
from jax import jit, lax, config
config.update("jax_platform_name", "cpu")  # "gpu"

def f1(i, x):
    y = jnp.zeros(x.shape)
    mask = jnp.full(x.shape, True)
    return jnp.where(mask, x, y) + 1

def f2(i, x):
    y = jnp.zeros(x.shape)
    mask = jnp.full(x.shape, True).astype(int)
    return x * mask + y * (1 - mask) + 1

@jit
def g1(x):
    return lax.fori_loop(0, 10000, f1, x)

@jit
def g2(x):
    return lax.fori_loop(0, 10000, f2, x)

for shape in [100, 1000, 10000, 100000]:
    x = jnp.arange(float(shape))
    g1(x)
    g2(x)
    print("Test for shape: {}".format(shape))
    %timeit z = g1(x).block_until_ready()
    %timeit z = g2(x).block_until_ready()
```
In CPU, `where` is faster
```
Test for shape: 100
227 µs ± 1.02 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
273 µs ± 7.21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Test for shape: 1000
932 µs ± 50 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.34 ms ± 17 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Test for shape: 10000
7.4 ms ± 29.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
10.5 ms ± 162 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Test for shape: 100000
72.9 ms ± 799 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
109 ms ± 586 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
In GPU, they are similar
```
Test for shape: 100
97.8 ms ± 632 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
97.7 ms ± 439 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Test for shape: 1000
97.2 ms ± 396 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
99.8 ms ± 3.06 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Test for shape: 10000
107 ms ± 383 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
107 ms ± 202 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
Test for shape: 100000
107 ms ± 346 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
104 ms ± 3.74 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
I expected the effect to be larger (and not in favor of `jnp.where`). Thanks for checking! I'll make the update to `jnp.where`.